### PR TITLE
feat: introduce an Api trait and its implementation

### PR DESCRIPTION
**Pull request by faassen**
The implementation is a thin wrapper over the RPC client.

This is exposed as a public API and will also be used to implement the CLI in iroh-ctl.

The idea is that the trait can be reimplemented to help test the iroh-ctl later, and it defines clearly what the public API is.

We smooth over the underlying client API somewhat: make `Cid` and `PeerId` into reference arguments, and use `&[]` slices instead of vectors for arguments as well.

In order to implement `ClientApi` I had to expose `P2pClient` and `StoreClient` in `iroh-rpc-client`.

### CODE REVIEW

File: iroh/src/lib.rs
Comment (fabricedesre) [inline iroh/src/lib.rs:8]: Making the client api depend on libp2p is problematic because of how much dependencies libp2p comes with. Something leaner and more self contained would be better.

Comment (faassen) [reply iroh/src/lib.rs:8]: I've brought this up to @dignifiedquire

**Review by ramfox — CHANGES_REQUESTED** [review]
very fun!

Some general notes:
- There are a few places with stray comments, I don't think I pointed them all out, but just something to be aware of as you are going back through the code
- speaking of comments, since this will be a public API, we are going to need documentation that can be generated for the docs.rs site for each public method/trait/enum/struct. I'd appreciate it if you could add `///` comments to any pieces you feel comfortable with. I'll help with this.

File: iroh/Cargo.toml
Comment (faassen) [reply iroh/Cargo.toml:18]: In each case here I used the procedure of looking at another crate with the same dependency and copying that over here.

That's not a policy but I think it's enough for this PR.

File: iroh/src/api.rs
Comment (faassen) [reply iroh/src/api.rs:64]: I don't understand this comment. Do you mean it changed from `addrs` to `peers`? I didn't change anything as this API is new, but it may definitely be inconsistent with the existing CLI somewhere. I saw both a `peers` command and an `addr` command in the CLI code if I recall correctly.

I think in general this whole API  needs a rethink the light of the discussion we had yesterday about the complex relationship between all this stuff.

File: iroh/src/api.rs
Comment (faassen) [reply iroh/src/api.rs:28]: Yes, I think that makes the most sense. I'll just remove it for now. Version makes more sense for the underlying nodes like p2p.

**Comment by faassen** [comment]
Thanks for the feedback! async clouded my brain making me over-async.

Concerning adding comments to produce public API documentation. I'm a bit wary of doing this early as I think we're still going to shake out a lot of details when we implement the CLI. I'd prefer to work on the CLI proposal first, let the API reflect that, and then do a comment pass in a future PR.

**Comment by faassen** [comment]
I've tried to integrate the `add` and `get`  code but bounced off both of them due to some async ``Send`` marker trait stuff.

The code in the CLI currently assumes that client is owned, not a reference like in my CLI code. But even once I fiddled with lifetimes and made client owned, I still get the problem.

So I think we need to look into this together.

**Comment by faassen** [comment]
I've tried to integrate `block_put` but I don't think I can. I'm pretty sure that `block_put` needs to take either byte data or perhaps some kind of file handle as the CLI command in kubo works with stdin. But `BlockCommands::Put` is based around `PathBuf`, and that doesn't seem right. We're not adding a file at all, just a block directly, right?

**Comment by faassen** [comment]
I try to look for opportunities to merge sooner rather than later. There are three issues left that I can determine:

* `add` and `get` are tricky; I've now prepared the API but I don't know how to integrate the code. I propose we do a later `add/get` PR where we figure this out.

* `block_put` as mentioned needs further discussion.

* As mentioned earlier, I want to hold off on API docs until more has been shaken out, in a future PR

None of these look like they have to block the merge, given the early state of this API in general.

File: iroh/src/api.rs
Comment (ramfox) [reply iroh/src/api.rs]: yes I did mean from `addrs` to `peers`, my bad. Yea I agree about the API rethink.

**Comment by faassen** [comment]
@dignifiedquire could you take a look at this? 

In particular @ramfox and I would like to know whether it's okay to use `#[async_trait(?Send)]` for the `GetAdd` trait. Without `?Send` it it's not possible to call the code in `getadd` without getting errors involving missing `Send` markers for `dyn std::future::Future<Output = std::result::Result<iroh_resolver::unixfs_builder::Directory, anyhow::Error>>`, `dyn libp2p::futures::Stream<Item = std::result::Result<(cid::CidGeneric<64_usize>, bytes::Bytes), anyhow::Error>>` and `(dyn tokio::io::AsyncRead + 'static)`, which puzzled us for a while.

The status of the various discussions:

* @ramfox and I agreed that API docs are going to land in a future PR once things shakes out

* the whole addresses API may need a rethink, also driven by what we want in the CLI. It's good enough to land for now as it mirrors the CLI.

File: iroh/src/api.rs
Comment (faassen) [inline iroh/src/api.rs:31]: @dignifiedquire here's the `?Send`

File: iroh/src/api.rs
Comment (dignifiedquire) [inline iroh/src/api.rs:46]: `Send` and `Sync` are auto imported, no need to add `std::marker::` infront of them

File: iroh/src/api.rs
Comment (dignifiedquire) [reply iroh/src/api.rs:31]: If we require `GetAdd` to be `Send`, we shouldn't use `?Send` here

**Comment by faassen** [comment]
We're going back to the drawing board. Closing this PR for the time being as I cook up something new.

**Pull request closed without merging**
